### ISSUE TITLE
Tree structure isn't always correct

### ISSUE THREAD

**Issue by itamarst** [jonathanj/eliottree#21]
Consider the following example:

```
4350e65c-bf31-4ecd-8d2d-b48a601dc6f6
    +-- check_links@1/started
        `-- url: http://localhost:8080
    +-- download@2,1/started
        `-- url: http://localhost:8080
        +-- download@2,2/succeeded
    +-- download@3,1/started
        `-- url: http://nosuchdomain
        +-- download@3,2/failed
            |-- exception: requests.exceptions.ConnectionError
            |-- reason: HTTPConnectionPool(host='nosuchdomain', port=80): Max retries exceeded with url: / (Caused by <class [...]
    +-- download@4,1/started
        `-- url: http://google.com
        +-- download@4,2/succeeded
    +-- check_links@5/succeeded
```

Notice that `download@2,1` and friends  are child actions of `check_links@1`, but that is not reflected in the structure.

**Comment by jonathanj** [comment]
The tree building code here is pretty ropey and definitely has some problems. Can you post the raw Eliot data this tree was generated from?

**Comment by jonathanj** [comment]
Hrm, from the new `task_level` docs:

> * `task_level`: The specific location of this message within the task's tree of actions.
>  For example, `[3, 2, 4]` indicates the message is the 4th child of the 2nd child of the 3rd child of the task.

Doesn't this mean that `download@2,1` is the 1st child of the 2nd child of the task, and since there is no `something@2`, that means it becomes a child of the task? If I understand correctly, in order for `download@2,1` to be a child action of `check_links@1` it would have to be named `download@1,x`.

**Comment by itamarst** [comment]
Well, notice how `@download3/2` is a child of `@download/3/1`? The messages for the former are probably `[3, 2, 1]` and `[3, 2, 2]`.

The problem with the current numbering system is that it's a bit tricksy. `[1]` is the *start* message of the top-level action. `[2]` is a child message of top-level action; `[2, 1]` is the start message of a child action of top-level action.

That prose description is therefore slightly inaccurate; message 1 isn't really child, it's start message.

**Comment by itamarst** [comment]
```
{"task_uuid":"e1bbc2e6-da49-4d29-8cfa-3e38aa1fc274","task_level":[1],"action_type":"check_links","url":"http:\/\/localhost:8080","timestamp":1438773463.2195188999,"action_status":"started"}
{"timestamp":1438773463.2197139263,"task_uuid":"e1bbc2e6-da49-4d29-8cfa-3e38aa1fc274","action_type":"extract_links","action_status":"started","task_level":[2,1]}
{"task_uuid":"e1bbc2e6-da49-4d29-8cfa-3e38aa1fc274","task_level":[2,2,1],"action_type":"download","url":"http:\/\/localhost:8080","timestamp":1438773463.219835043,"action_status":"started"}
{"timestamp":1438773463.2334361076,"task_uuid":"e1bbc2e6-da49-4d29-8cfa-3e38aa1fc274","action_type":"download","action_status":"succeeded","task_level":[2,2,2]}
{"task_uuid":"e1bbc2e6-da49-4d29-8cfa-3e38aa1fc274","task_level":[2,3],"action_type":"extract_links","timestamp":1438773463.2343270779,"found_urls":["http:\/\/nosuchdomain","http:\/\/google.com"],"action_status":"succeeded"}
{"task_uuid":"e1bbc2e6-da49-4d29-8cfa-3e38aa1fc274","task_level":[3,1],"action_type":"download","url":"http:\/\/nosuchdomain","timestamp":1438773463.2344739437,"action_status":"started"}
{"exception":"requests.exceptions.ConnectionError","task_level":[3,2],"action_type":"download","reason":"HTTPConnectionPool(host='nosuchdomain', port=80): Max retries exceeded with url: \/ (Caused by <class 'socket.gaierror'>: [Errno -2] Name or service not known)","timestamp":1438773463.2700130939,"task_uuid":"e1bbc2e6-da49-4d29-8cfa-3e38aa1fc274","action_status":"failed"}
{"task_uuid":"e1bbc2e6-da49-4d29-8cfa-3e38aa1fc274","task_level":[4,1],"action_type":"download","url":"http:\/\/google.com","timestamp":1438773463.2701640129,"action_status":"started"}
{"timestamp":1438773463.4309649467,"task_uuid":"e1bbc2e6-da49-4d29-8cfa-3e38aa1fc274","action_type":"download","action_status":"succeeded","task_level":[4,2]}
{"timestamp":1438773463.4311449528,"task_uuid":"e1bbc2e6-da49-4d29-8cfa-3e38aa1fc274","action_type":"check_links","action_status":"succeeded","task_level":[5]}
```

**Comment by itamarst** [comment]
The bug AFAICT is that top-level action gets tree structure wrong; all other actions seem fine. That is, top-level action in task doesn't realize all but first (action start) and last (action end) messages are children. I'll try to write minimal reproducer later today.

**Comment by itamarst** [comment]
My use case BTW is quickstart example for Eliot docs.

**Comment by jonathanj** [comment]
The task level in the tree is just the `task_level` key joined on commas. If `[1]` is a child of the top-level action and `[2]` is a sibling to `[1]` what is `[2,1]` when `[2]` is not present? The effective assumption eliottree makes here is that `[2,1]` becomes `[2]` not `[1,1]`.

It's a bit awkward to construct a tree given that the start and end messages are weird special cases that cannot be differentiated by the data alone. Is there a reliable method to detect the start and end messages for a task, bearing in mind that one can serialize a task and resume it in a different environment?

**Comment by itamarst** [comment]
Reproducer:
```
from eliot import to_file, start_action, Message
from sys import stdout
to_file(stdout)

with start_action(action_type="root_action"):
    Message.log(message_type="child_message")
    with start_action(action_type="child_action"):
        pass
```

Output:
```
d0f6ecc2-057e-4434-b353-0fd6439c7e3b
    +-- root_action@1/started
        `-- timestamp: 2015-08-05 09:17:52.063937
    +-- child_message@2/None
        |-- message_type: child_message
        `-- timestamp: 2015-08-05 09:17:52.064048
    +-- child_action@3,1/started
        `-- timestamp: 2015-08-05 09:17:52.064107
        +-- child_action@3,2/succeeded
            `-- timestamp: 2015-08-05 09:17:52.064149
    +-- root_action@4/succeeded
        `-- timestamp: 2015-08-05 09:17:52.064184
```

**Comment by itamarst** [comment]
Notice in above that if you shifted the two root_action messages to the left the structure would be correct.

**Comment by itamarst** [comment]
Serializing a task doesn't change where start and end messages happen; the remote messages in the second process are children. The way to detect start and end of action:

1. Start message is always number 1. So e.g. [1], [2, 1], [52, 353, 353, 1] are all start messages.
2. End message has action_type that matches message number 1 of that level. End messages that match the above might be [5], [2, 2], [52, 353, 353, 75] - they're always on same level 1.

So basically message number 1 is part of description of action, 2, 3, 4, 5, etc. are all children except for the last one that has matching level and is a succeeded/failed action message.

**Comment by itamarst** [comment]
If you have any ideas on how to improve the tree structure encoding that'd be cool too.

**Comment by jonathanj** [comment]
The logical way to describe the tree structure would surely be by `task_level` alone, which is what the [docstring of `TaskLevel.level`](https://github.com/ClusterHQ/eliot/blob/master/eliot/_action.py#L75) leads me to believe is the case, but apparently not?

To make sure my understanding of your earlier description is correct, my interpretation is:
 * `child_message@2` and `child_action@3,1` are ordered children of `root_action@1`
 * `child_action@3,2` is a child of `child_action@3,1`
 * `root_action@4` is a sibling of `root_action@1`

So if one were to describe the structure using `task_level` alone to produce a tree as per your description this would make the most sense to me:
 * `root_action@1` remains as is (the first node)
 * `child_message@2` becomes `child_message@1,1` (the first child of the first node)
 * `child_action@3,1` becomes `child_action@1,2` (the second child of the first node)
 * `child_action@3,2` becomes `child_action@1,2,1` (the first child of the second child of the first node)
 * `root_action@4` becomes `child_action@2` (the second node)

To produce the hypothetical output:
```
d0f6ecc2-057e-4434-b353-0fd6439c7e3b
    +-- root_action@1/started
        `-- timestamp: 2015-08-05 09:17:52.063937
        +-- child_message@1,1/None
            |-- message_type: child_message
            `-- timestamp: 2015-08-05 09:17:52.064048
        +-- child_action@1,2/started
            `-- timestamp: 2015-08-05 09:17:52.064107
            +-- child_action@1,2,1/succeeded
                `-- timestamp: 2015-08-05 09:17:52.064149
    +-- root_action@2/succeeded
        `-- timestamp: 2015-08-05 09:17:52.064184
```

**Comment by itamarst** [comment]
Hm. I think with that scheme you still can't differentiate between two unrelated messages and the start & finish messages of a childless action merely from the task level. On the other hand it does seem like you can figure out the *tree* structure just from task_level, so that is an improvement. Let me think about it a bit more.

**Comment by itamarst** [comment]
Oh, I notice BTW in your hypothetical output that success is rendered differently in root tree and child trees.

Random thought: 0 is always used for start message, MAXINT for action finish message?

**Comment by jonathanj** [comment]
@itamarst The rendering is simply matching what you suggested in your earlier comment: https://github.com/jonathanj/eliottree/issues/21#issuecomment-127994980

I think it's odd that in some cases success is a child (ala `[3,2]`) and some cases not (ala `[4]`).

**Comment by itamarst** [comment]
I feel like there's some miscommunication here...

**Comment by itamarst** [comment]
Ah, no, my fault, I hadn't noticed that typically eliot-tree visually renders end messages as children of start messages (even though that's not actual structure in eliot itself).
